hotrod-client-decoder/src/main/java/org/infinispan/hotrod/numeric/BranchlessParser.java

```diff
@@ -0,0 +1,133 @@
+   }
+
+   private static int readRawVarint40(ByteBuf buffer, int wholeOrMore) {
+      byte lastByte;
+      if (buffer.readableBytes() == 4 || (lastByte = buffer.getByte(buffer.readerIndex() + 4)) < 0) {
+         throw new CorruptedFrameException("malformed varint.");
+      }
+      buffer.skipBytes(5);
+      // add it to wholeOrMore
+      return wholeOrMore & 0x7F |
+            (((wholeOrMore >> 8) & 0x7F) << 7) |
+            (((wholeOrMore >> 16) & 0x7F) << 14) |
+            (((wholeOrMore >> 24) & 0x7F) << 21) |
+            (lastByte << 28);
+   }
+
+   public static long readRawVarint64(ByteBuf buffer) {
+      if (buffer.readableBytes() <= 4)
+         return readRawVarint32(buffer);
+
+      long wholeOrMore = buffer.getLongLE(buffer.readerIndex());
+      long firstOneOnStop = ~wholeOrMore & (0x80808080L << 32 | 0x80808080L);
+
+      // The value occupies the 9 bytes. We just unroll it and consume the bytes.
+      if (firstOneOnStop == 0) {
+         return readRawVarlong72(buffer, wholeOrMore);
+      }
+
+      // Consume the bytes containing the long.
+      int bitsToKeep = Long.numberOfTrailingZeros(firstOneOnStop) + 1;
+      buffer.skipBytes(bitsToKeep >> 3);
+
+      // Create a mask and create the continuation bytes for decoding.
+      long thisVarlongMask = firstOneOnStop ^ (firstOneOnStop - 1);
+      long continuation = wholeOrMore & thisVarlongMask;
+
+      // We parse it as 8 bytes, following the same state strategy for the integer parsing.
+      // Mix the continuation with bytes 1 and 3, bytes 2 and 4, bytes 5 and 7, bytes 6 and 8.
+      // It creates a 2 bit gap between each append of the 7 bits.
+      continuation = (continuation & 0x7F007F) |
+            ((continuation & 0x7F007F00) >> 1) |
+            ((continuation & (0x7F007F00L << 24))) |
+            ((continuation & (0x7F007F00L << 32)) >> 1);
+
+      // Now we isolate the bits in sequence. We check 14 bits at a time.
+      // The intervals are 0-14 bits, 16-30 (and shift 2), 32-46 (and shift 2 + 2), 48-62 (and shift 2 + 2 + 2).
+      return (continuation & 0x3FFF) |
+            ((continuation & 0x3FFF0000) >> 2) |
+            ((continuation & (0x3FFF0000L << 16)) >> 4) |
+            ((continuation & (0x3FFF0000L << 32)) >> 6);
+   }
+
+   private static long readRawVarlong72(ByteBuf buffer, long wholeOrMore) {
+      byte lastByte;
+      if (buffer.readableBytes() <= 8 || (lastByte = buffer.getByte(buffer.readerIndex() + 8)) < 0) {
+         throw new CorruptedFrameException("malformed varlong");
+      }
+      buffer.skipBytes(9);
+
+      // An unrolled version of the loop for decoding a varint.
+      // Shift right one byte at a time and shift left 7 bits at a time.
+      return wholeOrMore & 0x7F |
+            (((wholeOrMore >> 8) & 0x7F) << 7) |
+            (((wholeOrMore >> 16) & 0x7F) << 14) |
+            (((wholeOrMore >> 24) & 0x7F) << 21) |
```

Comment:
I'm wondering if it can be faster (and correct!) rewritten this way
```suggestion
            (((wholeOrMore) & 0x7F8) >> 1) |
            (((wholeOrMore) & 0x7F0) >> 2) |
            (((wholeOrMore) & 0x780) >> 3) |
            ...
```

Reply:
sorry I can't resist to this kind of puzzle :)

Reply:
mmm maybe the mask needs a fix

Reply:
I'll try updating it, let's see.